# Data Wrangling
---

## Data Acquisition

- Download sqlite file from Kaggle: https://www.kaggle.com/rtatman/188-million-us-wildfires
 - Found more recent data, up to 2018: https://www.fs.usda.gov/rds/archive/Catalog/RDS-2013-0009.5
- Convert to pandas dataframe:

In [2]:
# imports
import sqlite3
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

# create connection
cnx = sqlite3.connect('FPA_FOD_20210617.sqlite')
# get data
fires = pd.read_sql_query("SELECT * FROM Fires", cnx)

In [ ]:
# acquire function
def acquire_fires():
    '''
    This function returns a pandas dataframe of wildfire data from 1992-2018.
    '''
    # check for local csv
    import os
    if os.path.isfile('fires.csv'):
        fires = pd.read_csv('fires.csv', index_col=0)
    # if no local csv, get data from database
    else:
        # create connection
        cnx = sqlite3.connect('FPA_FOD_20210617.sqlite')
        # get data
        fires = pd.read_sql_query("SELECT * FROM Fires", cnx)
        # cache data to csv
        fires.to_csv('fires.csv')
    return fires

In [6]:
#Remove Limits On Viewing Dataframes
pd.set_option('display.max_columns', None)

# ignore pink warnings
import warnings
warnings.filterwarnings('ignore')

In [7]:
# save data as csv
fires.to_csv('fires.csv')

In [5]:
# view shape
fires.shape

(2166753, 37)

In [9]:
# preview data
fires.head()

FOD_ID      FPA_ID SOURCE_SYSTEM_TYPE SOURCE_SYSTEM NWCG_REPORTING_AGENCY  \
0       1  FS-1418826                FED   FS-FIRESTAT                    FS   
1       2  FS-1418827                FED   FS-FIRESTAT                    FS   
2       3  FS-1418835                FED   FS-FIRESTAT                    FS   
3       4  FS-1418845                FED   FS-FIRESTAT                    FS   
4       5  FS-1418847                FED   FS-FIRESTAT                    FS   

  NWCG_REPORTING_UNIT_ID  NWCG_REPORTING_UNIT_NAME SOURCE_REPORTING_UNIT  \
0                USCAPNF    Plumas National Forest                   511   
1                USCAENF  Eldorado National Forest                   503   
2                USCAENF  Eldorado National Forest                   503   
3                USCAENF  Eldorado National Forest                   503   
4                USCAENF  Eldorado National Forest                   503   

  SOURCE_REPORTING_UNIT_NAME LOCAL_FIRE_REPORT_ID LOCAL_INCIDENT_ID FIRE_CODE  \
0     Plumas National Forest                    1            PNF-47      BJ8K   
1   Eldorado National Forest                   13                13      AAC0   
2   Eldorado National Forest                   27               021      A32W   
3   Eldorado National Forest                   43                 6      None   
4   Eldorado National Forest                   44                 7      None   

  FIRE_NAME ICS_209_PLUS_INCIDENT_JOIN_ID ICS_209_PLUS_COMPLEX_JOIN_ID  \
0  FOUNTAIN                          None                         None   
1    PIGEON                          None                         None   
2     SLACK                          None                         None   
3      DEER                          None                         None   
4  STEVENOT                          None                         None   

  MTBS_ID MTBS_FIRE_NAME COMPLEX_NAME  FIRE_YEAR  DISCOVERY_DATE  \
0    None           None         None       2005   2/2/2005 0:00   
1    None           None         None       2004  5/12/2004 0:00   
2    None           None         None       2004  5/31/2004 0:00   
3    None           None         None       2004  6/28/2004 0:00   
4    None           None         None       2004  6/28/2004 0:00   

   DISCOVERY_DOY  DISCOVERY_TIME NWCG_CAUSE_CLASSIFICATION  \
0             33          1300.0                     Human   
1            133           845.0                   Natural   
2            152          1921.0                     Human   
3            180          1600.0                   Natural   
4            180          1600.0                   Natural   

                           NWCG_GENERAL_CAUSE NWCG_CAUSE_AGE_CATEGORY  \
0  Power generation/transmission/distribution                    None   
1                                     Natural                    None   
2                     Debris and open burning                    None   
3                                     Natural                    None   
4                                     Natural                    None   

        CONT_DATE  CONT_DOY  CONT_TIME  FIRE_SIZE FIRE_SIZE_CLASS   LATITUDE  \
0   2/2/2005 0:00      33.0     1730.0       0.10               A  40.036944   
1  5/12/2004 0:00     133.0     1530.0       0.25               A  38.933056   
2  5/31/2004 0:00     152.0     2024.0       0.10               A  38.984167   
3   7/3/2004 0:00     185.0     1400.0       0.10               A  38.559167   
4   7/3/2004 0:00     185.0     1200.0       0.10               A  38.559167   

    LONGITUDE       OWNER_DESCR STATE COUNTY FIPS_CODE         FIPS_NAME  
0 -121.005833              USFS    CA     63     06063     Plumas County  
1 -120.404444              USFS    CA     61     06061     Placer County  
2 -120.735556  STATE OR PRIVATE    CA     17     06017  El Dorado County  
3 -119.913333              USFS    CA      3     06003     Alpine County  
4 -119.933056              USFS    CA      3     06003  

In [4]:
# view dtypes
fires.dtypes

FOD_ID                             int64
FPA_ID                            object
SOURCE_SYSTEM_TYPE                object
SOURCE_SYSTEM                     object
NWCG_REPORTING_AGENCY             object
NWCG_REPORTING_UNIT_ID            object
NWCG_REPORTING_UNIT_NAME          object
SOURCE_REPORTING_UNIT             object
SOURCE_REPORTING_UNIT_NAME        object
LOCAL_FIRE_REPORT_ID              object
LOCAL_INCIDENT_ID                 object
FIRE_CODE                         object
FIRE_NAME                         object
ICS_209_PLUS_INCIDENT_JOIN_ID     object
ICS_209_PLUS_COMPLEX_JOIN_ID      object
MTBS_ID                           object
MTBS_FIRE_NAME                    object
COMPLEX_NAME                      object
FIRE_YEAR                          int64
DISCOVERY_DATE                    object
DISCOVERY_DOY                      int64
DISCOVERY_TIME                   float64
NWCG_CAUSE_CLASSIFICATION         object
NWCG_GENERAL_CAUSE                object
NWCG_CAUSE_AGE_C

In [8]:
# view nulls by column
fires.isna().sum()

FOD_ID                                 0
FPA_ID                                 0
SOURCE_SYSTEM_TYPE                     0
SOURCE_SYSTEM                          0
NWCG_REPORTING_AGENCY                  0
NWCG_REPORTING_UNIT_ID                 0
NWCG_REPORTING_UNIT_NAME               0
SOURCE_REPORTING_UNIT                  0
SOURCE_REPORTING_UNIT_NAME             0
LOCAL_FIRE_REPORT_ID             1701854
LOCAL_INCIDENT_ID                 734948
FIRE_CODE                        1797096
FIRE_NAME                         939607
ICS_209_PLUS_INCIDENT_JOIN_ID    2135993
ICS_209_PLUS_COMPLEX_JOIN_ID     2165833
MTBS_ID                          2153848
MTBS_FIRE_NAME                   2153848
COMPLEX_NAME                     2161081
FIRE_YEAR                              0
DISCOVERY_DATE                         0
DISCOVERY_DOY                          0
DISCOVERY_TIME                    754468
NWCG_CAUSE_CLASSIFICATION              1
NWCG_GENERAL_CAUSE                     0
NWCG_CAUSE_AGE_C

Dropping due to nulls:

LOCAL_FIRE_REPORT_ID             1701854
LOCAL_INCIDENT_ID                 734948
FIRE_CODE                        1797096
FIRE_NAME                         939607
ICS_209_PLUS_INCIDENT_JOIN_ID    2135993
ICS_209_PLUS_COMPLEX_JOIN_ID     2165833
MTBS_ID                          2153848
MTBS_FIRE_NAME                   2153848
COMPLEX_NAME                     2161081
DISCOVERY_TIME                    754468
NWCG_CAUSE_AGE_CATEGORY          2093127
CONT_TIME                         933151
COUNTY                            657235
FIPS_CODE                         657235
FIPS_NAME                         657236

In [10]:
# drop columns due to null values
fires.drop(columns=['LOCAL_FIRE_REPORT_ID',
                   'LOCAL_INCIDENT_ID',
                   'FIRE_CODE',
                   'FIRE_NAME',
                   'ICS_209_PLUS_INCIDENT_JOIN_ID',
                   'ICS_209_PLUS_COMPLEX_JOIN_ID',
                   'MTBS_ID',
                   'MTBS_FIRE_NAME',
                   'COMPLEX_NAME',
                   'DISCOVERY_TIME',
                   'NWCG_CAUSE_AGE_CATEGORY',
                   'CONT_TIME',
                   'COUNTY',
                   'FIPS_CODE',
                   'FIPS_NAME'], inplace=True)
fires.isna().sum()

FOD_ID                             0
FPA_ID                             0
SOURCE_SYSTEM_TYPE                 0
SOURCE_SYSTEM                      0
NWCG_REPORTING_AGENCY              0
NWCG_REPORTING_UNIT_ID             0
NWCG_REPORTING_UNIT_NAME           0
SOURCE_REPORTING_UNIT              0
SOURCE_REPORTING_UNIT_NAME         0
FIRE_YEAR                          0
DISCOVERY_DATE                     0
DISCOVERY_DOY                      0
NWCG_CAUSE_CLASSIFICATION          1
NWCG_GENERAL_CAUSE                 0
CONT_DATE                     854553
CONT_DOY                      854553
FIRE_SIZE                          0
FIRE_SIZE_CLASS                    0
LATITUDE                           0
LONGITUDE                          0
OWNER_DESCR                        0
STATE                              0
dtype: int64

In [14]:
# view rows where CONT_DATE is null
fires[fires.CONT_DATE.isna()]

FOD_ID               FPA_ID SOURCE_SYSTEM_TYPE SOURCE_SYSTEM  \
84              86           FS-1419068                FED   FS-FIRESTAT   
1506          1521           FS-1422198                FED   FS-FIRESTAT   
1507          1522           FS-1422199                FED   FS-FIRESTAT   
1552          1567           FS-1422292                FED   FS-FIRESTAT   
1555          1570           FS-1422297                FED   FS-FIRESTAT   
...            ...                  ...                ...           ...   
2166748  400482078  ICS209_2016_4363266          INTERAGCY     IA-ICS209   
2166749  400482080  ICS209_2016_4433939          INTERAGCY     IA-ICS209   
2166750  400482083  ICS209_2017_7335143          INTERAGCY     IA-ICS209   
2166751  400482085  ICS209_2017_7186124          INTERAGCY     IA-ICS209   
2166752  400482086  ICS209_2015_2836014          INTERAGCY     IA-ICS209   

        NWCG_REPORTING_AGENCY NWCG_REPORTING_UNIT_ID  \
84                         FS                USMTKNF   
1506                       FS                USMOMTF   
1507                       FS                USMOMTF   
1552                       FS                USMOMTF   
1555                       FS                USMOMTF   
...                       ...                    ...   
2166748                ST/C&L                USTNTNS   
2166749                ST/C&L                USALALS   
2166750                ST/C&L                USALALS   
2166751                    FS                USAROUF   
2166752                ST/C&L                USOKOKS   

               NWCG_REPORTING_UNIT_NAME SOURCE_REPORTING_UNIT  \
84             Kootenai National Forest                   114   
1506         Mark Twain National Forest                   905   
1507         Mark Twain National Forest                   905   
1552         Mark Twain National Forest                   905   
1555         Mark Twain National Forest                   905   
...                                 ...                   ...   
2166748  Tennessee Division of Forestry                 TNTNS   
2166749     Alabama Forestry Commission                 ALALS   
2166750     Alabama Forestry Commission                 ALALS   
2166751        Ouachita National Forest                 AROUF   
2166752   Oklahoma Division of Forestry                 OKECU   

                SOURCE_REPORTING_UNIT_NAME  FIRE_YEAR   DISCOVERY_DATE  \
84                Kootenai National Forest       2005    7/4/2005 0:00   
1506            Mark Twain National Forest       2005   3/11/2005 0:00   
1507            Mark Twain National Forest       2005   3/11/2005 0:00   
1552            Mark Twain National Forest       2005   3/15/2005 0:00   
1555            Mark Twain National Forest       2005   3/16/2005 0:00   
...                                    ...        ...              ...   
2166748     Tennessee Division of Forestry       2016  10/18/2016 0:00   
2166749        Alabama Forestry Commission       2016  10/22/2016 0:00   
2166750        Alabama Forestry Commission       2017    4/2/2017 0:00   
2166751           Ouachita National Forest       2017  11/30/2017 0:00   
2166752  East Central Area Dispatch Office       2015   7/20/2015 0:00   

         DISCOVERY_DOY NWCG_CAUSE_CLASSIFICATION  \
84                 185                     Human   
1506                70                     Human   
1507                70                     Human   
1552                74                     Human   
1555                75                     Human   
...                ...                       ...   
2166748            292                     Human   
2166749            296                     Human   
2166750             92                     Human   
2166751            334                     Human   
2166752            201                     Human   

                              NWCG_GENERAL_CAUSE CONT_DATE  CONT_DOY  \
84                                     Fireworks      None 

More unnecessary columns we have decided to drop:

FOD_ID                             0
FPA_ID                             0
SOURCE_SYSTEM_TYPE                 0
SOURCE_SYSTEM                      0
NWCG_REPORTING_AGENCY              0
NWCG_REPORTING_UNIT_ID             0
NWCG_REPORTING_UNIT_NAME           0
SOURCE_REPORTING_UNIT              0
SOURCE_REPORTING_UNIT_NAME         0
FIRE_SIZE_CLASS                    0
NWCG_CAUSE_CLASSIFICATION          1
CONT_DOY                      854553
DISCOVERY_DOY                      0
OWNER_DESCR

After our MVP, we may revisit some of these columns.

In [15]:
# drop columns not needed for MVP
fires.drop(columns=['FOD_ID',
                   'FPA_ID',
                   'SOURCE_SYSTEM_TYPE',
                   'SOURCE_SYSTEM',
                   'NWCG_REPORTING_AGENCY',
                   'NWCG_REPORTING_UNIT_ID',
                   'NWCG_REPORTING_UNIT_NAME',
                   'SOURCE_REPORTING_UNIT',
                   'SOURCE_REPORTING_UNIT_NAME',
                   'FIRE_SIZE_CLASS',
                   'NWCG_CAUSE_CLASSIFICATION',
                   'CONT_DOY',
                   'DISCOVERY_DOY',
                   'OWNER_DESCR'], inplace=True)

In [17]:
# view remaining columns
fires.columns

Index(['FIRE_YEAR', 'DISCOVERY_DATE', 'NWCG_GENERAL_CAUSE', 'CONT_DATE',
       'FIRE_SIZE', 'LATITUDE', 'LONGITUDE', 'STATE'],
      dtype='object')

In [18]:
# view remaining nulls
fires.isna().sum()

FIRE_YEAR                  0
DISCOVERY_DATE             0
NWCG_GENERAL_CAUSE         0
CONT_DATE             854553
FIRE_SIZE                  0
LATITUDE                   0
LONGITUDE                  0
STATE                      0
dtype: int64

In [16]:
# preview data
fires.head(2)

FIRE_YEAR  DISCOVERY_DATE                          NWCG_GENERAL_CAUSE  \
0       2005   2/2/2005 0:00  Power generation/transmission/distribution   
1       2004  5/12/2004 0:00                                     Natural   

        CONT_DATE  FIRE_SIZE   LATITUDE   LONGITUDE STATE  
0   2/2/2005 0:00       0.10  40.036944 -121.005833    CA  
1  5/12/2004 0:00       0.25  38.933056 -120.404444    CA

In [19]:
# view dtypes
fires.dtypes

FIRE_YEAR               int64
DISCOVERY_DATE         object
NWCG_GENERAL_CAUSE     object
CONT_DATE              object
FIRE_SIZE             float64
LATITUDE              float64
LONGITUDE             float64
STATE                  object
dtype: object

In [20]:
# view shape
fires.shape

(2166753, 8)

In [24]:
# view column names
fires.columns

Index(['FIRE_YEAR', 'DISCOVERY_DATE', 'NWCG_GENERAL_CAUSE', 'CONT_DATE',
       'FIRE_SIZE', 'LATITUDE', 'LONGITUDE', 'STATE'],
      dtype='object')

In [26]:
# change datatype
fires['DISCOVERY_DATE'] = pd.to_datetime(fires['DISCOVERY_DATE'])
fires.dtypes

FIRE_YEAR                      int64
DISCOVERY_DATE        datetime64[ns]
NWCG_GENERAL_CAUSE            object
CONT_DATE                     object
FIRE_SIZE                    float64
LATITUDE                     float64
LONGITUDE                    float64
STATE                         object
dtype: object

In [34]:
# change datatype
fires.CONT_DATE = pd.to_datetime(fires.CONT_DATE, errors='coerce')

In [35]:
# view changed dtypes
fires.dtypes

FIRE_YEAR                      int64
DISCOVERY_DATE        datetime64[ns]
NWCG_GENERAL_CAUSE            object
CONT_DATE             datetime64[ns]
FIRE_SIZE                    float64
LATITUDE                     float64
LONGITUDE                    float64
STATE                         object
dtype: object

In [37]:
# lowercase column names
fires.columns = fires.columns.str.lower()
fires.head(2)

fire_year discovery_date                          nwcg_general_cause  \
0       2005     2005-02-02  Power generation/transmission/distribution   
1       2004     2004-05-12                                     Natural   

   cont_date  fire_size   latitude   longitude state  
0 2005-02-02       0.10  40.036944 -121.005833    CA  
1 2004-05-12       0.25  38.933056 -120.404444    CA

In [42]:
# rename column name
fires.rename(columns={'nwcg_general_cause':'general_cause'}, inplace=True)

In [43]:
# preview data
fires.head(2)

fire_year discovery_date                               general_cause  \
0       2005     2005-02-02  Power generation/transmission/distribution   
1       2004     2004-05-12                                     Natural   

   cont_date  fire_size   latitude   longitude state  
0 2005-02-02       0.10  40.036944 -121.005833    CA  
1 2004-05-12       0.25  38.933056 -120.404444    CA

In [ ]:
# define prep function
def prep_fires(fires):
    '''
    This function takes in the fires dataframe and prepares it for our MVP
    exploration.
    '''
    # drop columns due to null values
    fires.drop(columns=['LOCAL_FIRE_REPORT_ID',
                   'LOCAL_INCIDENT_ID',
                   'FIRE_CODE',
                   'FIRE_NAME',
                   'ICS_209_PLUS_INCIDENT_JOIN_ID',
                   'ICS_209_PLUS_COMPLEX_JOIN_ID',
                   'MTBS_ID',
                   'MTBS_FIRE_NAME',
                   'COMPLEX_NAME',
                   'DISCOVERY_TIME',
                   'NWCG_CAUSE_AGE_CATEGORY',
                   'CONT_TIME',
                   'COUNTY',
                   'FIPS_CODE',
                   'FIPS_NAME'], inplace=True)
    # drop columns we don't need for MVP
    fires.drop(columns=['FOD_ID',
                   'FPA_ID',
                   'SOURCE_SYSTEM_TYPE',
                   'SOURCE_SYSTEM',
                   'NWCG_REPORTING_AGENCY',
                   'NWCG_REPORTING_UNIT_ID',
                   'NWCG_REPORTING_UNIT_NAME',
                   'SOURCE_REPORTING_UNIT',
                   'SOURCE_REPORTING_UNIT_NAME',
                   'FIRE_SIZE_CLASS',
                   'NWCG_CAUSE_CLASSIFICATION',
                   'CONT_DOY',
                   'DISCOVERY_DOY',
                   'OWNER_DESCR'], inplace=True)
    # convert columns to datetime
    fires['DISCOVERY_DATE'] = pd.to_datetime(fires['DISCOVERY_DATE'])
    fires.CONT_DATE = pd.to_datetime(fires.CONT_DATE, errors='coerce')
    # lowercase column names
    fires.columns = fires.columns.str.lower()
    # rename column name
    fires.rename(columns={'nwcg_general_cause':'general_cause'}, inplace=True)
    return fires